### PR TITLE
Phase2-hgx335J Try to add toplogy class for HGCal TB application

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Try to add toplogy class for HGCal TB application

#### PR validation:

Not fully tested. Need suggestion how to avoid type lookup error. Then add the topology builder and test it

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Nothing. special

### CODE REVIEW

**Comment by bsunanda** [comment]
@Dr15Jones, @makortel How can I avoid the type lookup error? I am not

**Comment by makortel** [comment]
> How can I avoid the type lookup error? I am not

@bsunanda Can you give more details of the error?

**Comment by bsunanda** [comment]
At the time of scram build it reports:

>> Building shared library tmp/slc7_amd64_gcc11/src/Geometry/CaloTopology/src/GeometryCaloTopology/libGeometryCaloTopology.so
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02769/slc7_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161/bin/../lib/gcc/x86_64-unknown-linux-gnu/11.2.1/../../../../x86_64-unknown-linux-gnu/bin/ld: tmp/slc7_amd64_gcc11/src/Geometry/CaloTopology/src/GeometryCaloTopology/HGCalTBTopology.cc.o: in function `std::type_info const& edm::typelookup::classTypeInfo<HGCalTBTopology>()':
HGCalTBTopology.cc:(.text+0x1f3): undefined reference to `typeinfo for HGCalTBTopology'
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02769/slc7_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161/bin/../lib/gcc/x86_64-unknown-linux-gnu/11.2.1/../../../../x86_64-unknown-linux-gnu/bin/ld: tmp/slc7_amd64_gcc11/src/Geometry/CaloTopology/src/GeometryCaloTopology/HGCalTBTopology.cc.o: in function `HGCalTBTopology::HGCalTBTopology(HGCalTBDDDConstants const&, int)':
HGCalTBTopology.cc:(.text+0x467): undefined reference to `vtable for HGCalTBTopology'
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02769/slc7_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161/bin/../lib/gcc/x86_64-unknown-linux-gnu/11.2.1/../../../../x86_64-unknown-linux-gnu/bin/ld: tmp/slc7_amd64_gcc11/src/Geometry/CaloTopology/src/GeometryCaloTopology/HGCalTBTopology.cc.o: in function `_GLOBAL__sub_I_HGCalTBTopology.cc':
HGCalTBTopology.cc:(.text.startup+0xee): undefined reference to `typeinfo for HGCalTBTopology'
collect2: error: ld returned 1 exit status
gmake: *** [config/SCRAM/GMake/Makefile.rules:1717: tmp/slc7_amd64_gcc11/src/Geometry/CaloTopology/src/GeometryCaloTopology/libGeometryCaloTopology.so] Error 1
gmake: *** Waiting for unfinished jobs....
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2


________________________________

**Comment by bsunanda** [comment]
@makortel @Dr15Jones I have added what message I get during building the library. Please take a look at the previous entry in his PR

**Comment by makortel** [comment]
@cmsbuild, please test

Hmm, not immediately clear. Let's see if it reproduces here.

File: Geometry/CaloTopology/interface/HGCalTBTopology.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+#include <vector>
+#include <iostream>
+
+class HGCalTBTopology : public CaloSubdetectorTopology {
+public:
+  /// create a new Topology
+  HGCalTBTopology(const HGCalTBDDDConstants& hdcons, int subdet);
+
+  /// virtual destructor
+  ~HGCalTBTopology() override = default;
```
Comment (Dr15Jones) [inline Geometry/CaloTopology/interface/HGCalTBTopology.h:18]: Try removing this line and see if that helps the linking problem.

Comment (bsunanda) [reply Geometry/CaloTopology/interface/HGCalTBTopology.h:18]: That didn't help either

**Comment by Dr15Jones** [comment]
One more idea, in the header have
```C++
~HGCalTBTopology() override;
```

Then in the .cc file do
```C++
HGCalTBTopology::~HGCalTBTopology() {}
```

That will force the vtable to be generated in the .cc file which should give us the typeinfo as well.

**Comment by makortel** [comment]
> One more idea, in the header have
> 
> ```c++
> ~HGCalTBTopology() override;
> ```
> 
> Then in the .cc file do
> 
> ```c++
> HGCalTBTopology::~HGCalTBTopology() {}
> ```
> 
> That will force the vtable to be generated in the .cc file which should give us the typeinfo as well.

I actually tried that in parallel :D It works to get beyond the complaing of missing vtable, after which the linker complains
```
undefined reference to `HGCalTBTopology::detId2denseId(DetId const&) const'
```
and indeed the definition of that overridden virtual function is missing. That _may_ be the root cause for the link error.

File: Geometry/CaloTopology/interface/HGCalTBTopology.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  DecodedDetId geomDenseId2decId(const uint32_t& hi) const;
+  DecodedDetId decode(const DetId& id) const;
+  DetId encode(const DecodedDetId& id_) const;
+
+  DetId::Detector detector() const { return det_; }
+  ForwardSubdetector subDetector() const { return subdet_; }
+  bool detectorType() const { return false; }
+
+private:
+  HGCalTBTopology() : hdcons_(nullptr) {}
```
Comment (makortel) [inline Geometry/CaloTopology/interface/HGCalTBTopology.h:115]: If nothing uses this constructor, why define it?

**Comment by bsunanda** [comment]
@makortel @Dr15Jones Thanks a lot. It is built at last

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
+geometry

**Comment by perrotta** [comment]
+1

**Pull request merged**